pkg/controller/machineset/controller.go

```diff
@@ -99,45 +136,60 @@ func (c *MachineSetControllerImpl) Get(namespace, name string) (*v1alpha1.Machin
 }
 
 // syncReplicas essentially scales machine resources up and down.
-func (c *MachineSetControllerImpl) syncReplicas(machineSet *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
-	// Take ownership of machines if not already owned.
-	for _, machine := range machines {
-		if shouldAdopt(machineSet, machine) {
-			c.adoptOrphan(machineSet, machine)
-		}
-	}
-
-	var result error
-	currentMachineCount := int32(len(machines))
-	desiredReplicas := *machineSet.Spec.Replicas
-	diff := int(currentMachineCount - desiredReplicas)
-
+func (c *MachineSetControllerImpl) syncReplicas(ms *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
+	diff := len(machines) - int(*(ms.Spec.Replicas))
```

Comment:
added wrapper to check for nil and default to 0

Reply:
This doesn't seem like the right approach. The point of the value being potentially nil is that it is optional. We shouldn't treat it as a "0" value when it is not present. 

Also -- it would be best to to see how the number of times we need to check for a nil value can be reduced. It seems like there must be a way to reduce the number of checks / places where the wrapper function is invoked.

Reply:
This should never be nil at this point in the system, as the validation for machineset objects will default a nil value to 1.
https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/apis/cluster/v1alpha1/machineset_types.go#L53
https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/apis/cluster/v1alpha1/machineset_types.go#L163

Treating it as a 0 here is just to avoid working with a nil value.

Reply:
Here's my thoughts. I'd rather see this reverted to how you had it previously or proper handling of nil values introduced (perhaps as a pre-processing check so the nil checks do not need to be all over the case).

1. Defaulting to a zero value adds uncertainty and questions about correctness to the code. Will it ever happen? It seems like it won't right now given the implementation in machineset_types.go but when that implementation changes the behavior of syncReplicas will be altered unknowingly.
2. If the type is intended to be optional we should treat it as optional or just punt on it and fault when we encounter a nil value. 

So my argument is that either we should properly handle the nil value or revert to the previous behavior where panics would occur when a nil value is encountered (so remove the wrapper). The current state is nebulous and unclear. Sorry for the back and forth here.

Reply:
This should be addressed now.